Null Pointer Exception when importing gpx file

**Describe the bug**
When importing gpx files through settings -> import all tracks, I get a null pointer exception.
Unfortunately I can't attach the file due to Copyright issues.
![Screenshot_20210302-183554_OpenTracks](https://user-images.githubusercontent.com/24672655/109694916-99722f80-7b8b-11eb-9f86-ec85d53d9354.png)


**Technical information**
 - OS: LineageOS 15.1

**Logcat**

```
<<< log_count = 21 >>>
[03-02 18:39:47.876 1395:1428 I/Adreno-EGL]
<qeglDrvAPI_eglInitialize:379>: EGL 1.4 QUALCOMM build: Nondeterministic_AU_msm8974_LA.BF.1.1.3_RB1__release_AU (I741a3d36ca)
OpenGL ES Shader Compiler Version: E031.29.00.00
Build Date: 04/04/16 Mon
Local Branch: mybranch19053788
Remote Branch: quic/LA.BF.1.1.3_rb1.12
Local Patches: NONE
Reconstruct Branch: NOTHING

[03-02 18:39:47.876 1395:1428 I/zygote]
android::hardware::configstore::V1_0::ISurfaceFlingerConfigs::hasWideColorDisplay retrieved: 0

[03-02 18:39:47.876 1395:1428 I/OpenGLRenderer]
Initialized EGL, version 1.4

[03-02 18:39:47.876 1395:1428 D/OpenGLRenderer]
Swap behavior 1

[03-02 18:39:47.884 1395:1428 W/Adreno-EGL]
<qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE

[03-02 18:39:47.887 1395:1428 W/Adreno-EGL]
<qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE

[03-02 18:39:47.897 1395:1428 D/vndksupport]
Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.

[03-02 18:39:47.922 1395:1428 W/Adreno-EGL]
<qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE

[03-02 18:39:47.934 1395:1428 D/vndksupport]
Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.

[03-02 18:39:49.543 1395:1428 W/Adreno-EGL]
<qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE

[03-02 18:39:49.551 1395:1428 D/vndksupport]
Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.

[03-02 18:39:49.645 1395:1428 D/OpenGLRenderer]
endAllActiveAnimators on 0x99be8000 (RippleDrawable) with handle 0x99c0b400

[03-02 18:39:51.592 1395:1428 D/OpenGLRenderer]
endAllActiveAnimators on 0x8a7cbd00 (RippleDrawable) with handle 0x99c0bf30

[03-02 18:39:53.958 1395:1428 W/Adreno-EGL]
<qeglDrvAPI_eglGetConfigAttrib:607>: EGL_BAD_ATTRIBUTE

[03-02 18:39:53.969 1395:1428 D/vndksupport]
Loading /vendor/lib/hw/gralloc.msm8974.so from current namespace instead of sphal namespace.

[03-02 18:39:54.044 1395:9280 W/AbstractFileTrackImporter]
Marker without time ignored.

[03-02 18:39:54.051 1395:9280 I/chatty]
uid=10161(de.dennisguse.opentracks) AsyncTask #13 identical 11 lines

[03-02 18:39:54.052 1395:9280 W/AbstractFileTrackImporter]
Marker without time ignored.

[03-02 18:39:54.053 1395:9280 E/StringUtils]
Invalid XML dateTime value

[03-02 18:39:54.066 1395:9280 E/AbstractFileTrackImporter]
Unable to import file
java.lang.NullPointerException: text
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.createTrackPoint(AbstractFileTrackImporter.java:436)
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.getTrackPoint(AbstractFileTrackImporter.java:339)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.onTrackPointEnd(GpxFileTrackImporter.java:205)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.endElement(GpxFileTrackImporter.java:112)
	at org.apache.harmony.xml.ExpatParser.endElement(ExpatParser.java:156)
	at org.apache.harmony.xml.ExpatParser.appendBytes(Native Method)
	at org.apache.harmony.xml.ExpatParser.parseFragment(ExpatParser.java:513)
	at org.apache.harmony.xml.ExpatParser.parseDocument(ExpatParser.java:474)
	at org.apache.harmony.xml.ExpatReader.parse(ExpatReader.java:316)
	at org.apache.harmony.xml.ExpatReader.parse(ExpatReader.java:279)
	at javax.xml.parsers.SAXParser.parse(SAXParser.java:390)
	at javax.xml.parsers.SAXParser.parse(SAXParser.java:187)
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.importFile(AbstractFileTrackImporter.java:123)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.importFile(GpxFileTrackImporter.java:39)
	at de.dennisguse.opentracks.io.file.importer.ImportService.importFile(ImportService.java:65)
	at de.dennisguse.opentracks.io.file.importer.ImportService.onHandleWork(ImportService.java:45)
	at androidx.core.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:392)
	at androidx.core.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:383)
	at android.os.AsyncTask$2.call(AsyncTask.java:333)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)

[03-02 18:39:54.068 1395:9280 D/ImportService]
Parser error: java.lang.NullPointerException: text
de.dennisguse.opentracks.io.file.importer.ImportParserException: java.lang.NullPointerException: text
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.importFile(AbstractFileTrackImporter.java:130)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.importFile(GpxFileTrackImporter.java:39)
	at de.dennisguse.opentracks.io.file.importer.ImportService.importFile(ImportService.java:65)
	at de.dennisguse.opentracks.io.file.importer.ImportService.onHandleWork(ImportService.java:45)
	at androidx.core.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:392)
	at androidx.core.app.JobIntentService$CommandProcessor.doInBackground(JobIntentService.java:383)
	at android.os.AsyncTask$2.call(AsyncTask.java:333)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.NullPointerException: text
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.createTrackPoint(AbstractFileTrackImporter.java:436)
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.getTrackPoint(AbstractFileTrackImporter.java:339)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.onTrackPointEnd(GpxFileTrackImporter.java:205)
	at de.dennisguse.opentracks.io.file.importer.GpxFileTrackImporter.endElement(GpxFileTrackImporter.java:112)
	at org.apache.harmony.xml.ExpatParser.endElement(ExpatParser.java:156)
	at org.apache.harmony.xml.ExpatParser.appendBytes(Native Method)
	at org.apache.harmony.xml.ExpatParser.parseFragment(ExpatParser.java:513)
	at org.apache.harmony.xml.ExpatParser.parseDocument(ExpatParser.java:474)
	at org.apache.harmony.xml.ExpatReader.parse(ExpatReader.java:316)
	at org.apache.harmony.xml.ExpatReader.parse(ExpatReader.java:279)
	at javax.xml.parsers.SAXParser.parse(SAXParser.java:390)
	at javax.xml.parsers.SAXParser.parse(SAXParser.java:187)
	at de.dennisguse.opentracks.io.file.importer.AbstractFileTrackImporter.importFile(AbstractFileTrackImporter.java:123)
	... 10 more

```

--- a/src/androidTest/java/de/dennisguse/opentracks/util/StringUtilsTest.java
+++ b/src/androidTest/java/de/dennisguse/opentracks/util/StringUtilsTest.java
@@ -94,18 +94,12 @@ public class StringUtilsTest {
         assertEquals("<![CDATA[hello]]]]><![CDATA[>there]]>", StringUtils.formatCData("hello]]>there"));
     }
 
-    /**
-     * Tests {@link StringUtils#parseTime(String)} with fractional seconds.
-     */
     @Test
     public void testGetTime_fractional() {
         assertGetTime("2010-05-04T03:02:01.352Z", 2010, 5, 4, 3, 2, 1, 352);
         assertGetTime("2010-05-04T03:02:01.3529Z", 2010, 5, 4, 3, 2, 1, 352);
     }
 
-    /**
-     * Tests {@link StringUtils#parseTime(String)} with time zone.
-     */
     @Test
     public void testGetTime_timezone() {
         assertGetTime("2010-05-04T03:02:01", 2010, 5, 4, 3, 2, 1, 0);
@@ -118,9 +112,6 @@ public class StringUtilsTest {
         assertGetTime("2010-05-04T03:02:01-05:00", 2010, 5, 4, 8, 2, 1, 0);
     }
 
-    /**
-     * Tests {@link StringUtils#parseTime(String)} with fractional seconds and time zone.
-     */
     @Test
     public void testGetTime_fractionalAndTimezone() {
         assertGetTime("2010-05-04T03:02:01.352Z", 2010, 5, 4, 3, 2, 1, 352);

--- a/src/main/java/de/dennisguse/opentracks/io/file/importer/AbstractFileTrackImporter.java
+++ b/src/main/java/de/dennisguse/opentracks/io/file/importer/AbstractFileTrackImporter.java
@@ -21,6 +21,8 @@ import android.database.sqlite.SQLiteConstraintException;
 import android.net.Uri;
 import android.util.Log;
 
+import androidx.annotation.NonNull;
+
 import org.xml.sax.Locator;
 import org.xml.sax.SAXException;
 import org.xml.sax.helpers.DefaultHandler;
@@ -122,7 +124,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
         try {
             SAXParserFactory.newInstance().newSAXParser().parse(inputStream, this);
             return trackIds.get(0);
-        } catch (IOException | SAXException | ParserConfigurationException e) {
+        } catch (IOException | SAXException | ParserConfigurationException | ParsingException e) {
             Log.e(TAG, "Unable to import file", e);
             if (trackIds.size() > 0) {
                 cleanImport();
@@ -304,7 +306,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
         }
     }
 
-    protected void addMarker() throws SAXException {
+    protected void addMarker() throws ParsingException {
         // Markers must have a time, else cannot match to the track points
         if (time == null) {
             Log.w(TAG, "Marker without time ignored.");
@@ -335,7 +337,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
         markers.add(marker);
     }
 
-    protected TrackPoint getTrackPoint() throws SAXException {
+    protected TrackPoint getTrackPoint() throws ParsingException {
         TrackPoint trackPoint = createTrackPoint();
 
         if (trackPoint.hasLocation()) {
@@ -350,7 +352,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
                 trackPoint.setTime(time);
 
             } else if (!LocationUtils.isValidLocation(trackPoint.getLocation())) {
-                throw new SAXException(createErrorMessage("Invalid location detected: " + trackPoint));
+                throw new ParsingException(createErrorMessage("Invalid location detected: " + trackPoint));
             }
         }
 
@@ -424,16 +426,13 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
         return null;
     }
 
-    /**
-     * Creates a location.
-     */
-    private TrackPoint createTrackPoint() throws SAXException {
+    private TrackPoint createTrackPoint() throws ParsingException {
         TrackPoint trackPoint = new TrackPoint(TrackPoint.Type.TRACKPOINT);
 
         try {
             trackPoint.setTime(StringUtils.parseTime(time));
         } catch (Exception e) {
-            throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse time: %s", time)), e);
+            throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse time: %s", time)), e);
         }
 
         if (latitude == null || longitude == null) {
@@ -444,7 +443,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
             trackPoint.setLatitude(Double.parseDouble(latitude));
             trackPoint.setLongitude(Double.parseDouble(longitude));
         } catch (NumberFormatException e) {
-            throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse latitude longitude: %s %s", latitude, longitude)), e);
+            throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse latitude longitude: %s %s", latitude, longitude)), e);
         }
 
 
@@ -452,7 +451,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
             try {
                 trackPoint.setAltitude(Double.parseDouble(altitude));
             } catch (NumberFormatException e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse altitude: %s", altitude)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse altitude: %s", altitude)), e);
             }
         }
 
@@ -460,14 +459,14 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
             try {
                 trackPoint.setSpeed(Float.parseFloat(speed));
             } catch (Exception e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse speed: %s", speed)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse speed: %s", speed)), e);
             }
         }
         if (heartrate != null) {
             try {
                 trackPoint.setHeartRate_bpm(Float.parseFloat(heartrate));
             } catch (Exception e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse heart rate: %s", heartrate)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse heart rate: %s", heartrate)), e);
             }
         }
 
@@ -475,7 +474,7 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
             try {
                 trackPoint.setCyclingCadence_rpm(Float.parseFloat(cadence));
             } catch (Exception e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse cadence: %s", cadence)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse cadence: %s", cadence)), e);
             }
         }
 
@@ -483,14 +482,14 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
             try {
                 trackPoint.setElevationGain(Float.parseFloat(gain));
             } catch (Exception e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse elevation gain: %s", gain)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse elevation gain: %s", gain)), e);
             }
         }
         if (loss != null) {
             try {
                 trackPoint.setElevationLoss(Float.parseFloat(loss));
             } catch (Exception e) {
-                throw new SAXException(createErrorMessage(String.format(Locale.US, "Unable to parse elevation loss: %s", loss)), e);
+                throw new ParsingException(createErrorMessage(String.format(Locale.US, "Unable to parse elevation loss: %s", loss)), e);
             }
         }
 
@@ -545,5 +544,22 @@ abstract class AbstractFileTrackImporter extends DefaultHandler implements Track
 
         // The buffered locations
         final List<TrackPoint> bufferedTrackPoints = new ArrayList<>();
+    }
+
+    public static class ParsingException extends RuntimeException {
+
+        private ParsingException(@NonNull String message) {
+            super(message);
+        }
+
+        private ParsingException(@NonNull String message, Exception cause) {
+            super(message, cause);
+        }
+
+        @NonNull
+        @Override
+        public String toString() {
+            return "" + getMessage();
+        }
     }
 }

--- a/src/main/java/de/dennisguse/opentracks/io/file/importer/ImportService.java
+++ b/src/main/java/de/dennisguse/opentracks/io/file/importer/ImportService.java
@@ -57,33 +57,33 @@ public class ImportService extends JobIntentService {
             trackImporter = new KmzTrackImporter(this, file.getUri());
         } else {
             Log.d(TAG, "Unsupported file format.");
-            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file.getName(), getString(R.string.import_unsupported_format));
+            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file, getString(R.string.import_unsupported_format));
             return;
         }
 
         try (InputStream inputStream = getContentResolver().openInputStream(file.getUri())) {
             Track.Id trackId = trackImporter.importFile(inputStream);
             if (trackId != null) {
-                sendResult(ImportServiceResultReceiver.RESULT_CODE_IMPORTED, trackId, file.getName(), getString(R.string.import_file_imported, file.getName()));
+                sendResult(ImportServiceResultReceiver.RESULT_CODE_IMPORTED, trackId, file, getString(R.string.import_file_imported, file.getName()));
             } else {
-                sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, trackId, file.getName(), getString(R.string.import_unable_to_import_file, file.getName()));
+                sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, trackId, file, getString(R.string.import_unable_to_import_file, file.getName()));
             }
         } catch (IOException e) {
             Log.d(TAG, "Unable to import file", e);
-            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file.getName(), getString(R.string.import_unable_to_import_file, e.getMessage()));
+            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file, getString(R.string.import_unable_to_import_file, e.getMessage()));
         } catch (ImportParserException e) {
             Log.d(TAG, "Parser error: " + e.getMessage(), e);
-            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file.getName(), getString(R.string.import_parser_error, e.getMessage()));
+            sendResult(ImportServiceResultReceiver.RESULT_CODE_ERROR, null, file, getString(R.string.import_parser_error, e.getMessage()));
         } catch (ImportAlreadyExistsException e) {
             Log.d(TAG, "Track already exists: " + e.getMessage(), e);
-            sendResult(ImportServiceResultReceiver.RESULT_CODE_ALREADY_EXISTS, null, file.getName(), e.getMessage());
+            sendResult(ImportServiceResultReceiver.RESULT_CODE_ALREADY_EXISTS, null, file, e.getMessage());
         }
     }
 
-    private void sendResult(int resultCode, Track.Id trackId, String fileName, String message) {
+    private void sendResult(int resultCode, Track.Id trackId, DocumentFile file, String message) {
         Bundle bundle = new Bundle();
         bundle.putParcelable(ImportServiceResultReceiver.RESULT_EXTRA_TRACK_ID, trackId);
-        bundle.putString(ImportServiceResultReceiver.RESULT_EXTRA_FILENAME, fileName);
+        bundle.putString(ImportServiceResultReceiver.RESULT_EXTRA_FILENAME, file.getName());
         bundle.putString(ImportServiceResultReceiver.RESULT_EXTRA_MESSAGE, message);
         resultReceiver.send(resultCode, bundle);
     }

--- a/src/main/java/de/dennisguse/opentracks/io/file/importer/ImportViewModel.java
+++ b/src/main/java/de/dennisguse/opentracks/io/file/importer/ImportViewModel.java
@@ -95,7 +95,7 @@ public class ImportViewModel extends AndroidViewModel implements ImportServiceRe
         private int successCount;
         private int existsCount;
         private int errorCount;
-        private ArrayList<Track.Id> importedTrackIds = new ArrayList<>();
+        private final ArrayList<Track.Id> importedTrackIds = new ArrayList<>();
         private final ArrayList<String> fileErrors = new ArrayList<>();
 
         public int getTotalCount() {

